contracts/utils/Strings.sol

```diff
@@ -70,16 +71,35 @@ library Strings {
      * @dev Converts a `uint256` to its ASCII `string` hexadecimal representation.
      */
     function toHexString(uint256 value) internal pure returns (string memory) {
-        if (value == 0) {
-            return "0x00";
-        }
-        uint256 temp = value;
-        uint256 length = 0;
-        while (temp != 0) {
-            length++;
-            temp >>= 8;
+        unchecked {
+            uint256 length = 1;
+
+            // compute log256(value), and add it to length
+            uint256 valueCopy = value;
+            if (valueCopy >> 128 > 0) {
```

Comment:
`if (valueCopy > 1 << 128) {` will be equivalent, but should be cheaper, the shifting will be evaluated at compilation time.

Reply:
I'm not sure the shift is done at compile time. Can you confirm ?

Reply:
As of 0.8.13, yes, I've changed some constants to `1 << X` and the gas usage hasn't changed at all.

Reply:
>but should be cheaper

Why? I'm seeing the compiler emit code sequences that cost exactly the same for the two expressions.

`PUSH1 0x1 PUSH1 0x80 SHL DUP2 LT`
vs
`PUSH1 0x80 DUP2 SWAP1 SHR ISZERO`

Reply:
>did you enable the optimizer?

Yes. For constants Solidity will often prioritize code size before runtime cost.